recover.go

```diff
@@ -17,48 +17,292 @@ package connect
+		}
+	}
+}
+
+type recoverStreamRequest struct {
+	StreamingHandlerConn
+	msg any
+}
+
+func (r *recoverStreamRequest) Any() any {
+	return r.msg
+}
+
+func (r *recoverStreamRequest) Header() http.Header {
+	return r.RequestHeader()
+}
+
+func (r *recoverStreamRequest) HTTPMethod() string {
+	return http.MethodPost // streams always use POST
+}
+
+func (r *recoverStreamRequest) internalOnly() {
+}
+
+func (r *recoverStreamRequest) setRequestMethod(_ string) {
+	// only invoked internally for unary RPCs; safe to ignore
+}
+
+type recoverStreamingHandlerConn struct {
+	StreamingHandlerConn
+	req atomic.Pointer[any]
+}
+
+func (r *recoverStreamingHandlerConn) Receive(msg any) error {
+	err := r.StreamingHandlerConn.Receive(msg)
+	if err == nil {
+		// Note: The framework instantiates msg, passes it to
+		// this method, and then returns it to the application.
+		// It is possible that the application could mutate the
+		// value, so what we provide to the recover handler would
+		// then differ from the message actually received. But
+		// this is no different than if the RPC handler mutated
+		// the request message for a unary RPC and interceptors
+		// later examined it via Request.Any. So we tolerate the
+		// possibility for server-stream requests, too.
+		r.req.Store(&msg)
 	}
+	return err
+}
+
+type emptyRequest Spec
+
+func (e emptyRequest) Any() any {
+	return nil
+}
+
+func (e emptyRequest) Spec() Spec {
+	return Spec(e)
+}
+
+func (e emptyRequest) Peer() Peer {
+	return Peer{}
+}
+
+func (e emptyRequest) Header() http.Header {
+	return nil
+}
+
+func (e emptyRequest) HTTPMethod() string {
+	return http.MethodPost
+}
+
+func (e emptyRequest) internalOnly() {
+}
+
+func (e emptyRequest) setRequestMethod(_ string) {
+	// only invoked internally for unary RPCs; safe to ignore
+}
+
+type errStreamingClientConn struct {
+	spec Spec
+	err  error
+}
+
+func (e *errStreamingClientConn) Spec() Spec {
+	return e.spec
+}
+
+func (e *errStreamingClientConn) Peer() Peer {
+	return Peer{}
+}
+
+func (e *errStreamingClientConn) Send(_ any) error {
+	return e.err
+}
+
+func (e *errStreamingClientConn) RequestHeader() http.Header {
+	// Clients can add headers before calling Send, so this must be mutable/non-nil.
+	return http.Header{} // TODO: memoize so we never allocate more than one?
+}
+
+func (e *errStreamingClientConn) CloseRequest() error {
+	return e.err
+}
+
+func (e *errStreamingClientConn) Receive(_ any) error {
+	return e.err
+}
+
+func (e *errStreamingClientConn) ResponseHeader() http.Header {
+	return nil
+}
+
+func (e *errStreamingClientConn) ResponseTrailer() http.Header {
+	return nil
+}
+
+func (e *errStreamingClientConn) CloseResponse() error {
+	return e.err
+}
+
+type recoverStreamingClientConn struct {
+	StreamingClientConn
+
+	//nolint:containedctx // must memoize the stream context to pass to recover handler
+	ctx    context.Context
+	handle func(context.Context, AnyRequest, any) error
+	req    atomic.Pointer[any]
```

Comment:
`atomic.Pointer` I think should be just `atomic.Value` here as you are using `any` anyway.